data-prepper-core/src/main/java/org/opensearch/dataprepper/parser/CircuitBreakingBuffer.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.parser;
+
+import org.opensearch.dataprepper.breaker.CircuitBreaker;
+import org.opensearch.dataprepper.model.CheckpointState;
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.record.Record;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.TimeoutException;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Decorator for {@link Buffer} which checks a {@link CircuitBreaker}
+ * before writing records.
+ *
+ * @param <T> The type of record.
+ * @since 2.1
+ */
+class CircuitBreakingBuffer<T extends Record<?>> implements Buffer<T> {
+    private final Buffer<T> buffer;
+    private final CircuitBreaker circuitBreaker;
+
+    /**
+     * Constructor
+     *
+     * @param buffer The inner buffer which is being decorated
+     * @param circuitBreaker The circuit breaker to check
+     */
+    public CircuitBreakingBuffer(final Buffer<T> buffer, final CircuitBreaker circuitBreaker) {
+        this.buffer = requireNonNull(buffer);
+        this.circuitBreaker = requireNonNull(circuitBreaker);
+    }
+
+    @Override
+    public void write(final T record, final int timeoutInMillis) throws TimeoutException {
+        checkBreaker();
```

Comment:
Do we expect any performance impact from checking the breakers?

Reply:
I did some testing of the call to check memory usage and it was sub-milliseconds.

I did add a locking mechanism since we have multiple threads checking at the same time. That might have an impact.

An alternative approach I considered was to schedule a task (say every 1 second) to check the memory usage and update the breaker value rather than have one thread handle it. Then this `checkBreaker` is really just reading a boolean value.

Reply:
Having a dedicated task checking the memory seems like a good idea. Do you see any downsides to that approach?

Reply:
The extra thread may use some CPU, but that is probably not too much of an issue.

I'll probably add no synchronization in this case since the exact timing isn't too important (some threads may get circuit breaker before others). This should remove any performance impact from synchronization. Let me know if you think otherwise.